interfaces/prompting/requestrules/requestrules.go

```diff
@@ -452,31 +463,57 @@ func (rdb *RuleDB) addRulePermissionToTree(rule *Rule, permission string) []prom
 	var conflicts []prompting_errors.RuleConflict
 
 	addVariant := func(index int, variant patterns.PatternVariant) {
+		variantStr := variant.String()
+		existingEntry, exists := permVariants.VariantEntries[variantStr]
+		if !exists {
+			newVariantEntries[variantStr] = variantEntry{
+				Variant: variant,
+				Outcome: rule.Outcome,
+				RuleIDs: map[prompting.IDType]bool{rule.ID: true},
+			}
+			return
+		}
+		if existingEntry.Outcome != rule.Outcome {
+			// Outcomes mismatch, so if there are any unexpired rules, these
+			// all conflict with the new rule. If all rules are expired, this
+			// is fine, and we can replace the existing entry with a new one.
+			for id := range existingEntry.RuleIDs {
+				if rdb.isRuleWithIDExpired(id, rule.Timestamp) {
+					expiredRules[id] = true
+					continue
+				}
+				// Conflicting non-expired rule
+				conflicts = append(conflicts, prompting_errors.RuleConflict{
+					Permission:    permission,
+					Variant:       variantStr,
+					ConflictingID: id.String(),
+				})
+			}
+		}
+		if len(conflicts) > 0 {
```

Comment:
would it make sense to check this early in the fuction as well then?

Reply:
The motivation for waiting until this point in the closure is so we can include every conflict with every variant. The `RuleConflictError` wraps these in `addRuleToTree`, so the error response has all the information for the client. I think if we checked any earlier, we would miss conflicts.